### PR TITLE
Add `Lockable` to Base, to bundle a lock with its resource

### LINKED ISSUE #52897: Add `Lockable` to Base?
e.g. like
```julia
struct Lockable{T, L <: Base.AbstractLock}
    value::T
    lock::L
end
Lockable(value) = Lockable(value, ReentrantLock())
Base.getindex(l::Lockable) = l.value
```
as in https://github.com/JuliaServices/ConcurrentUtilities.jl/blob/main/src/lockable.jl.

I think this could be ergonomic for use in Base since there's various bits of global state that each need a lock (e.g. https://github.com/JuliaLang/julia/issues/49778), many of which are of the form
```julia
const RESOURCE = ...
const RESOURCE_LOCK = ReentrantLock()
```
and one does `@lock(RESOURCE_LOCK, ...access RESOURCE...)`. With `Lockable`, they could be
```julia
const RESOURCE = Lockable(...)
```
and used by `@lock(RESOURCE, ...access RESOURCE[]...)`.

By grouping the resource and it's lock into a struct, it makes it slightly harder to accidentally use the resource without the lock (since one needs to index into it), and it makes it crystal clear what lock is for what resource.

### PR DESCRIPTION
**Pull request by ericphanson**
redo of https://github.com/JuliaLang/julia/pull/34400
closes #52897 

I think this can be useful, so putting it back up. (First commit I copied from https://github.com/JuliaServices/ConcurrentUtilities.jl/blob/main/src/lockable.jl, so I added @quinnj as a co-author, then I realized it came from #34400, so I added @DilumAluthge).

I am not sure about the `lock(f, ::Lockable)` method: it is nice because it unpacks the value for you, but it is weird because it unpacks the value for you. For a `Lockable`, `f` must accept one argument, whereas for a `Lock`, `f` must be 0-arg. A `Lockable` is not `<:AbstractLock` here, so maybe this is allowed, but if we deleted this `lock` method, we could inherit from `AbstractLock` and just use the generic one (requiring folks to unpack the value within the locked region themselves, as is the case for `@lock`). I think it is preferred these days to use `@lock` anyway, so having the `lock(f,::Lockable)` method may be of limited value anyway.

I searched Base and came up with two places that could currently use this internally, `TEMP_CLEANUP` and `env_dict`, so I've added commits to show the usage there.

### CODE REVIEW

File: base/lock.jl [hunk 1]
```diff
@@ -290,6 +290,66 @@ macro lock_nofail(l, expr)
...
+ 1
+```
+"""
+struct Lockable{T, L <: Base.AbstractLock}
+    value::T
+    lock::L
+end
+
+Lockable(value) = Lockable(value, ReentrantLock())
+Base.getindex(l::Lockable) = l.value
```
Comment (vtjnash) [inline base/lock.jl:325]:
```suggestion
Base.getindex(l::Lockable) = (assert_havelock(l); l.value)
```

File: base/lock.jl [hunk 2]
```diff
@@ -290,6 +290,66 @@ macro lock_nofail(l, expr)
...
+
+"""
+function Base.lock(f, l::Lockable)
+    lock(l.lock) do
+        f(l.value)
+    end
+end
+
+# implement the rest of the Lock interface on Lockable
+Base.islocked(l::Lockable) = islocked(l.lock)
```
Comment (vtjnash) [inline base/lock.jl:348]:
```suggestion
```
This one is not really recommended to use, and the need for it is gone

**Review by vtjnash — COMMENTED** [review]
SGTM

File: base/lock.jl (hunk 1 above)
Comment (ericphanson) [reply base/lock.jl:325]: ooh, I didn't know `assert_havelock` existed, nice!

File: NEWS.md [hunk 3]
```diff
@@ -60,6 +60,8 @@ difference between defining a `main` function and executing the code directly at
 Multi-threading changes
 -----------------------
 
+* There is a new struct `Lockable{T, L<:AbstractLock}` that makes it easy to bundle a resource and its lock together.
```
Comment (vchuravy) [inline NEWS.md:63]: Reference to this PR?

Comment (ericphanson) [reply NEWS.md:63]:
```suggestion
* A new exported struct `Lockable{T, L<:AbstractLock}` makes it easy to bundle a resource and its lock together ([#52898]).
```

**Comment by ericphanson** [comment]
ArgTools tests are failing here, should be fixed by https://github.com/JuliaIO/ArgTools.jl/pull/35

File: base/lock.jl [hunk 4]
```diff
@@ -290,6 +290,64 @@ macro lock_nofail(l, expr)
...
+julia> lock(summary, locked_list)
+"1-element Vector{Int64}"
+```
+"""
+struct Lockable{T, L <: AbstractLock}
+    value::T
+    lock::L
+end
+
+Lockable(value) = Lockable(value, ReentrantLock())
```
Comment (grandinj) [inline base/lock.jl:323]: drive-by comment: 
- I don't know the history here
-  ReentrantLock is much safer than a non-reentrantlock for general use
-  ReentrantLock is significantly more expensive than a non-reentrant lock (like a factor 10 worse when it is heavily used)
- most code does not need a ReentrantLock.

Comment (KristofferC) [reply base/lock.jl:323]: Might be better to not provide a default here?

Comment (vtjnash) [reply base/lock.jl:323]: On the contrary, ReentrantLock is slightly faster than non-reentrant lock

Comment (gbaraldi) [reply base/lock.jl:323]: If the user knows that the access is fast and not reentrant then they can just specify the SpinLock.

Comment (vtjnash) [reply base/lock.jl:323]: They should never do that though, since it is a worse sort of lock

Comment (grandinj) [reply base/lock.jl:323]: > On the contrary, ReentrantLock is slightly faster than non-reentrant lock

That is interesting. I assumed that Julia was similar to C++ where std::mutex is considerably cheaper than std::recursive_mutex.

Comment (vtjnash) [reply base/lock.jl:323]: Any reference for that claim? It seems odd that C++ would implement std::recursive_mutex so badly, if they already have a good std::mutex. For contrast, note that Apple implements their standard mutex on top of their recursive_mutex (see https://opensource.apple.com/source/Libc/Libc-262/pthreads/pthread_mutex.c.auto.html), so it would be pretty unexpected for this to be badly very implemented as you claim

Comment (grandinj) [reply base/lock.jl:323]: > Any reference for that claim?

Oh dear, I'm going to need a brown paper bag. My apologies.

I ran some tests (linked here: https://gist.github.com/grandinj/d2ba5525bdccea1b72825b5ff6028648)

And the difference appears to be fairly minor - No real difference on Windows, and std::mutex is only 30% faster on Linux.

**Comment by vtjnash** [comment]
Would probably be best not to hold up this PR for that change, but is almost done now, so can just see about rebasing onto soon: https://github.com/JuliaLang/julia/pull/53124

**Comment by inkydragon** [comment]
Test failed related:

```
Error in testset Pkg:
Error During Test at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XG3Q6T6R70.0/build/default-honeycrisp-XG3Q6T6R70-0/julialang/julia-master/julia-7f8a4a525c/share/julia/test/testdefs.jl:24
  Got exception outside of a @test
  LoadError: MethodError: no method matching empty!(::Lockable{Dict{String, Bool}, ReentrantLock})
  Closest candidates are:
    ...
  Stacktrace:
    [1] top-level scope
      @ /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XG3Q6T6R70.0/build/default-honeycrisp-XG3Q6T6R70-0/julialang/julia-master/julia-7f8a4a525c/share/julia/stdlib/v1.11/Pkg/test/runtests.jl:96
```

**Comment by ericphanson** [comment]
Urg, I regret making the temp file changes here, I just wanted to show `Lockable` had uses within Base. At least I checked if `TEMP_FILE isa Lockable` in the ArgTools PR so it's not too late to revert that here.

**Comment by ericphanson** [comment]
Failure seems unrelated now:

```julia
cmdlineargs                                       (7) |         failed at 2024-02-04T00:12:08.906
Test Failed at /cache/build/tester-amdci4-12/julialang/julia-master/julia-8a35ded8ae/share/julia/test/cmdlineargs.jl:1065
  Expression: success(pipeline(setenv(`$(Base.julia_cmd()) --bug-report=rr-local -e 'exit()'`, "JULIA_RR_RECORD_ARGS" => "-n --nested=ignore", "_RR_TRACE_DIR" => temp_trace_dir); ))
```

**Comment by vtjnash** [comment]
Seems odd it would fail repeatedly though, when that test hasn't usually been a problem

**Comment by vtjnash** [comment]
Indeed it is related. Running locally gives:
```
WARNING: both ConcurrentUtilities and Base export "Lockable"; uses of it in module Connections must be qualified
```
```
ERROR: LoadError: UndefVarError: `Lockable` not defined in `HTTP.Connections`
Stacktrace:
 [1] top-level scope
   @ ~/.julia/packages/HTTP/bDoga/src/Connections.jl:385
 [2] include(mod::Module, _path::String)
   @ Base ./Base.jl:556
 [3] include(x::String)
   @ HTTP ~/.julia/packages/HTTP/bDoga/src/HTTP.jl:1
 [4] top-level scope
   @ ~/.julia/packages/HTTP/bDoga/src/HTTP.jl:40
 [5] include
   @ ./Base.jl:556 [inlined]
 [6] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::String)
   @ Base ./loading.jl:2427
 [7] top-level scope
   @ stdin:3
in expression starting at /home/vtjnash/.julia/packages/HTTP/bDoga/src/Connections.jl:1
in expression starting at /home/vtjnash/.julia/packages/HTTP/bDoga/src/HTTP.jl:1
in expression starting at stdin:3
```

**Comment by DilumAluthge** [comment]
Sounds like HTTP.jl has a `using ConcurrentUtilities` somewhere that they need to change to `using ConcurrentUtilities: Lockable`.

**Comment by vtjnash** [comment]
Or `ConcurrentUtilities` should add a case `if isdefined(Base, :Lockable); using Base: Lockable; else; include("lockable.jl"); end`

**Comment by vtjnash** [comment]
https://github.com/JuliaServices/ConcurrentUtilities.jl/pull/31

**Comment by adienes** [comment]
would be cool to have this in 1.11 if possible

**Comment by vtjnash** [comment]
I don't have the ability to make a new https://github.com/JuliaServices/ConcurrentUtilities.jl/pull/31 release, but once that is done, this should be able to merge

File: base/exports.jl [hunk 5]
```diff
@@ -702,6 +702,7 @@ export
     @lock,
     notify,
     ReentrantLock,
+    Lockable,
```
Comment (vtjnash) [inline base/exports.jl:705]:
```suggestion
```

File: test/misc.jl [hunk 6]
```diff
@@ -129,6 +129,34 @@ let l = ReentrantLock()
     @test_throws ErrorException unlock(l)
 end
 
+# Lockable{T, L<:AbstractLock}
+let
```
Comment (vtjnash) [inline test/misc.jl:133]:
```suggestion
using Base.Lockable
let
    @test_broken Base.isexported(Base, :Lockable)
```

**Comment by vtjnash** [comment]
You know what though, we don't actually have to wait for that. We will put off doing the actual export for a later day though.

**Comment by quinnj** [comment]
submitted a new ConcurrentUtilities.jl release

**Comment by vtjnash** [comment]
Thanks! We will still merge this today, and then now can merge the export in a couple days or weeks to give people time to update to the new release

**Comment by ericphanson** [comment]
Thanks for getting this in @vtjnash!
